File: Divinity/Apple/Sources/OmnideaCore/Contacts.swift
import COmnideaFFI
import Foundation

/// Swift wrapper around the Rust Contacts (module registry).
public final class Contacts: @unchecked Sendable {
    private let ptr: OpaquePointer

    public init() {
        ptr = divi_contacts_new()!
    }

    deinit {
        divi_contacts_free(ptr)
    }

    /// Register a module from a JSON-encodable value.
    public func register(_ info: some Encodable) throws {
        let json = try JSONEncoder().encode(info)
        let jsonString = String(data: json, encoding: .utf8)!

        let result = divi_contacts_register(ptr, jsonString)
        if result != 0 {
            try OmnideaError.check()
            throw OmnideaError(message: "Contacts register failed")
        }
    }

    /// Unregister a module by ID. Dependents are shut down first.
    public func unregister(_ moduleId: String) throws {
        let result = divi_contacts_unregister(ptr, moduleId)
        if result != 0 {
            try OmnideaError.check()
            throw OmnideaError(message: "Contacts unregister failed")
        }
    }

    /// Shut down all modules in dependency order.
    public func shutdownAll() {
        divi_contacts_shutdown_all(ptr)
    }

    /// Look up a module by ID. Returns JSON string or nil.
    public func lookup(_ moduleId: String) -> String? {
        guard let json = divi_contacts_lookup(ptr, moduleId) else { return nil }
        defer { divi_free_string(json) }
        return String(cString: json)
    }

    /// Get all registered modules as a JSON array string.
    public func allModulesJSON() -> String {
        guard let json = divi_contacts_all_modules(ptr) else { return "[]" }
        defer { divi_free_string(json) }
        return String(cString: json)
    }

    /// Get all registered module IDs.
    public func registeredModuleIds() -> [String] {
        guard let json = divi_contacts_registered_module_ids(ptr) else { return [] }
        defer { divi_free_string(json) }
        let data = Data(String(cString: json).utf8)
        return (try? JSONDecoder().decode([String].self, from: data)) ?? []
    }

    /// Get modules that depend on the given module, as JSON.
    public func dependentsOf(_ moduleId: String) -> String {
        guard let json = divi_contacts_dependents_of(ptr, moduleId) else { return "[]" }
        defer { divi_free_string(json) }
        return String(cString: json)
    }
}

File: Divinity/Android/src/main/kotlin/com/omnidea/core/Contacts.kt
package com.omnidea.core

/** Callback interface for module shutdown. */
fun interface ShutdownCallback {
    fun onShutdown()
}

/** Module registry with dependency tracking. */
class Contacts : AutoCloseable {
    private var ptr: Long

    init {
        ptr = nativeNew()
        check(ptr != 0L) { "Failed to create Contacts" }
    }

    /** Register a module from a JSON string. Throws on error. */
    fun register(json: String) {
        checkOpen()
        OmnideaError.check(nativeRegister(ptr, json))
    }

    /** Register a module with a shutdown callback. */
    fun registerWithShutdown(json: String, callback: ShutdownCallback) {
        checkOpen()
        OmnideaError.check(nativeRegisterWithShutdown(ptr, json, callback))
    }

    /** Unregister a module by ID. Throws on error. */
    fun unregister(moduleId: String) {
        checkOpen()
        OmnideaError.check(nativeUnregister(ptr, moduleId))
    }

    /** Shut down all modules in dependency order. */
    fun shutdownAll() {
        checkOpen()
        nativeShutdownAll(ptr)
    }

    /** Look up a module by ID. Returns JSON string or null. */
    fun lookup(moduleId: String): String? {
        checkOpen()
        return nativeLookup(ptr, moduleId)
    }

    /** Get all registered modules as a JSON array string. */
    fun allModulesJson(): String? {
        checkOpen()
        return nativeAllModules(ptr)
    }

    /** Get all registered module IDs as a JSON array string. */
    fun registeredModuleIdsJson(): String? {
        checkOpen()
        return nativeRegisteredModuleIds(ptr)
    }

    /** Get all dependents of a module as a JSON array string. */
    fun dependentsOf(moduleId: String): String? {
        checkOpen()
        return nativeDependentsOf(ptr, moduleId)
    }

    override fun close() {
        if (ptr != 0L) {
            nativeFree(ptr)
            ptr = 0L
        }
    }

    private fun checkOpen() = check(ptr != 0L) { "Contacts already closed" }

    companion object {
        init { NativeLoader.ensureLoaded() }

        @JvmStatic private external fun nativeNew(): Long
        @JvmStatic private external fun nativeFree(ptr: Long)
        @JvmStatic private external fun nativeRegister(ptr: Long, json: String): Int
        @JvmStatic private external fun nativeRegisterWithShutdown(ptr: Long, json: String, callback: ShutdownCallback): Int
        @JvmStatic private external fun nativeUnregister(ptr: Long, moduleId: String): Int
        @JvmStatic private external fun nativeShutdownAll(ptr: Long)
        @JvmStatic private external fun nativeLookup(ptr: Long, moduleId: String): String?
        @JvmStatic private external fun nativeAllModules(ptr: Long): String?
        @JvmStatic private external fun nativeRegisteredModuleIds(ptr: Long): String?
        @JvmStatic private external fun nativeDependentsOf(ptr: Long, moduleId: String): String?
    }
}
